subprojects/docs/src/docs/userguide/running-builds/configuration_cache.adoc

```diff
@@ -359,119 +580,149 @@ For example if you need a `Logger` in your task actions you should use `Task.log
-// - p2 is using `project.exec {}`
-// - in task types
-// - in tasks declared in scripts
-//   https://gradle.slack.com/archives/CH43P3A10/p1591477979141600
+Here's an example of a problematic task type using the `Project` object at execution time:
+
+====
+include::sample[dir="snippets/configurationCache/projectAtExecution/groovy",files="build.gradle[tags=task-type]"]
+include::sample[dir="snippets/configurationCache/projectAtExecution/kotlin",files="build.gradle.kts[tags=task-type]"]
+====
+<1> this will be reported as a problem because the task action is using the `Project` object at execution time
+
+The following is how it should be done instead:
+
+====
+include::sample[dir="snippets/configurationCache/projectAtExecutionFixed/groovy",files="build.gradle[tags=task-type]"]
+include::sample[dir="snippets/configurationCache/projectAtExecutionFixed/kotlin",files="build.gradle.kts[tags=task-type]"]
+====
+<1> no more problem reported, the injected `ExecOperations` service is supported as a replacement for `project.exec {}`
+
+In the same vein, if you encounter the same problem with an ad-hoc task declared in a script as follows:
+
+====
+include::sample[dir="snippets/configurationCache/projectAtExecution/groovy",files="build.gradle[tags=ad-hoc-task]"]
+include::sample[dir="snippets/configurationCache/projectAtExecution/kotlin",files="build.gradle.kts[tags=ad-hoc-task]"]
+====
+<1> this will be reported as a problem because the task action is using the `Project` object at execution time
+
+Here's how the task declaration above can be fixed:
+
+====
+include::sample[dir="snippets/configurationCache/projectAtExecutionFixed/groovy",files="build.gradle[tags=ad-hoc-task]"]
+include::sample[dir="snippets/configurationCache/projectAtExecutionFixed/kotlin",files="build.gradle.kts[tags=ad-hoc-task]"]
+====
+<1> services can't be injected directly in scripts, we need an extra type to convey the injection point
+<2> create an instance of the extra type using `project.object` outside the task action
+<3> no more problem reported, the task action references `injected` that provides the `ExecOperations` service, supported as a replacement for `project.exec {}`
+
+As you can see above, fixing ad-hoc tasks declared in scripts requires quite a bit of ceremony.
+It is a good time to think about extracting your task declaration as a proper task class as shown previously.
+
+The following table shows what APIs or injected service should be used as a replacement for each of the `Project` methods.
 
 [cols="a,a",options="header"]
 |===
 | Instead of:
 | Use:
 
 | `project.rootDir`
-| A task field or script variable to capture the root directory.
+| A task property or script variable to capture the root directory.
 
 | `project.projectDir`
-| `link:{javadocPath}/org/gradle/api/file/ProjectLayout.html[ProjectLayout].projectDirectory`
+| link:{javadocPath}/org/gradle/api/file/ProjectLayout.html#getProjectDirectory--[ProjectLayout.projectDirectory]
```

Comment:
I wouldn't suggest people use this in task action, as this usually makes the project directory an implicit input of the action. Instead, the task should have a property to carry this input.

We should really disallow `ProjectLayout` to be used from task actions.